pkg/flow/internal/controller/config.go

```diff
@@ -34,3 +37,166 @@ func NewConfigNode(block *ast.BlockStmt, globals ComponentGlobals, isInModule bo
 		return nil, diags
 	}
 }
+
+// ConfigNodeMap represents the config BlockNodes in their explicit types.
+// This is helpful when validating node conditions specific to config node
+// types.
+type ConfigNodeMap struct {
+	logging     *LoggingConfigNode
+	tracing     *TracingConfigNode
+	argumentMap map[string]*ArgumentConfigNode
+	exportMap   map[string]*ExportConfigNode
+}
+
+// NewConfigNodeMap will create an initial ConfigNodeMap. Append must be called
+// to populate NewConfigNodeMap.
+func NewConfigNodeMap() *ConfigNodeMap {
+	return &ConfigNodeMap{
+		logging:     nil,
+		tracing:     nil,
+		argumentMap: map[string]*ArgumentConfigNode{},
+		exportMap:   map[string]*ExportConfigNode{},
+	}
+}
+
+// Append will add a config node to the ConfigNodeMap unless it already exists
+// in which case it will return a diag specifying the problem.
+func (nodeMap *ConfigNodeMap) Append(configNode BlockNode) diag.Diagnostics {
```

Comment:
I can. It looks like it will panic in current state and confirmed via local testing. I might want to improve this as part of this PR if we go that route.

https://github.com/grafana/agent/blob/main/pkg/flow/internal/dag/dag.go#L66

Reply:
You can also do a Get before adding it to avoid changing the Add implementation